### PR TITLE
Fix translationKeyExists to work with onError handler

### PR DESCRIPTION
**Pull request by mgingras**
## Description
With custom onError handler in i18n `translate` doesn't raise an exception causing `translationKeyExists` to always return true.

Also considered testing if the resulting string from translate was empty AND we weren't using the default on error. i.e.

```ts
translationKeyExists(id: string) {
  try {
    const translation = this.translate(id);
    return translation !== '' || this.onError !== defaultOnError;
  } catch (error) {
    return false;
  }
}
```

## Type of change

- [x] `react-i18n` Patch: Bug/ Documentation fix (non-breaking change which fixes an issue or adds documentation)

## Checklist

- [x] I have added a changelog entry, prefixed by the type of change noted above

### CODE REVIEW

**Comment by CameronGorrie** [comment]
There was [some discussion](https://shopify.slack.com/archives/CCNRSKLTH/p1570716399032000) that pointed to removing this utility altogether

**Comment by mgingras** [comment]
Interesting, @CameronGorrie . The case we're trying to handle at the moment is where we have translations for a subset of things, and in the case we don't have a translation want to fall back to a default. Not sure how we'd handle it right now without the ability to check if a translation exists, and trying to translate blindly isn't great because it always throws an error.

File: packages/react-i18n/src/i18n.ts [hunk 1]
```diff
@@ -204,11 +204,15 @@ export class I18n {
   }
 
   translationKeyExists(id: string) {
+    const errorHandler = this.onError;
     try {
+      (this.onError as any) = defaultOnError;
```
Comment (lemonmade) [inline packages/react-i18n/src/i18n.ts]: I know this works but it feels a little icky to me. I'd rather do a proper check for whether the key exists rather than relying on the exception for control flow, especially now that it requires weird mutations of self

Comment (CameronGorrie) [inline packages/react-i18n/src/i18n.ts]: This property should use the `private` modifier instead of being casted

Comment (mgingras) [reply packages/react-i18n/src/i18n.ts]: How do you do that? :)

Comment (mgingras) [reply packages/react-i18n/src/i18n.ts]: Tried again, lmk what you think. 😄

Comment (CameronGorrie) [reply packages/react-i18n/src/i18n.ts]: I think Chris and I submitted our comments around the same time so my point isn't valid now, but the reason you had to cast the property is because `onError` was declared using the `readonly` modifier so you couldn't reassign this member without tricking the compiler. Changing the modifier to `private` would have allowed reassignment without the cast. [More info](https://www.typescriptlang.org/docs/handbook/classes.html#public-private-and-protected-modifiers)

**Comment by CameronGorrie** [comment]
I made a mistake porting this utility directly from the legacy translation library in web by not thinking too much about why it was being used. Perhaps there are cases when this utility is necessary but I'm thinking there's a solution to what you're trying to do.

> The case we're trying to handle at the moment is where we have translations for a subset of things, and in the case we don't have a translation want to fall back to a default

Would a function containing a dictionary of translations, a lookup, and optional fallback work? Maybe something like this:

```ts
function getTranslation(subset: string) {
  const dictionary = {
    subsetOne: i18n.translate('MyComponent.stringOne'),
    subsetTwo: i18n.translate('MyComponent.stringTwo'),
    default: i18n.translate('MyComponent.default'),
  }

  return dictionary[subset] || dictionary.default;
}
```

File: packages/react-i18n/src/i18n.ts (hunk 1 above)
Comment (mgingras) [reply packages/react-i18n/src/i18n.ts]: Cool, thanks for the details

**Comment by mgingras** [comment]
That would work, however then we're tracking what's defined both in the translation as well as in the business logic.

**Comment by mgingras** [comment]
@CameronGorrie @lemonmade any further thoughts, as it stands this fn doesn't work, if this fix isn't the right approach, I'm happy to try something else, or we could remove it

**Comment by CameronGorrie** [comment]
My vote would be to remove it. There's only a single usage in Web so I would assume other apps are not using it heavily

**Comment by alxclark** [comment]
👋 We are currently using this fn for the Checkout Settings page in the following context 👇 

https://github.com/Shopify/web/blob/master/app/sections/Checkout/CheckoutSettings/components/CheckoutLanguage/CheckoutLanguage.tsx#L30-L31

We had to introduce this since calling `i18n.translate` with a key that doesn't exist results in the following error.

image

I am not sure that defining a `getTranslation` fn for all the locale in our context would be best since adding new translations would imply updating everytime a new locale is added.
